### PR TITLE
Fix cuFFT callback compilations - v2

### LINKED ISSUE #4746: Find a way to cleanly support fft callbacks
Comment: https://github.com/cupy/cupy/pull/4734#issuecomment-785392652

### PR DESCRIPTION
**Pull request by leofang**
Close #4746.

This is an alternative of #4743 (proposal v2 mentioned there). The solution here is to avoid cimporting any CuPy modules so that we don't expose Cython-level APIs. The incurred overhead is OK ~~(which I will post later)~~: about +/- 5 us (yes, it can be faster in some cases, which I didn't expect), see https://github.com/cupy/cupy/pull/4853#issuecomment-794903196.

TODO:
- [x] Fix https://github.com/cupy/cupy/issues/4746#issuecomment-785543235

### CODE REVIEW

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Before:
```
full fftn
get_fft_plan        :    CPU:  153.272 us   +/- 3.355 (min:  150.945 / max:  166.278) us     GPU-0:  159.038 us   +/- 4.217 (min:  156.320 / max:  175.904) us
fftn                :    CPU:   33.019 us   +/- 1.499 (min:   31.115 / max:   37.133) us     GPU-0: 7434.614 us   +/-51.791 (min: 7341.824 / max: 7526.272) us

stack fftn
get_fft_plan        :    CPU:  159.775 us   +/-39.466 (min:  147.214 / max:  331.048) us     GPU-0:  165.499 us   +/-39.701 (min:  152.320 / max:  337.728) us
fftn                :    CPU:   74.661 us   +/- 8.435 (min:   68.006 / max:   96.624) us     GPU-0:17072.426 us   +/-35.179 (min:16983.295 / max:17122.145) us

stack fft
get_fft_plan        :    CPU:  150.631 us   +/- 2.059 (min:  148.672 / max:  156.752) us     GPU-0:  156.584 us   +/- 2.300 (min:  154.400 / max:  163.264) us
fft                 :    CPU:   16.575 us   +/- 0.795 (min:   15.735 / max:   18.899) us     GPU-0: 2076.304 us   +/- 3.245 (min: 2071.968 / max: 2083.648) us
```

After:
```
full fftn
get_fft_plan        :    CPU:  155.049 us   +/- 3.409 (min:  152.386 / max:  164.402) us     GPU-0:  161.062 us   +/- 4.085 (min:  158.080 / max:  173.952) us
fftn                :    CPU:   32.376 us   +/- 1.328 (min:   30.627 / max:   36.655) us     GPU-0: 7438.229 us   +/-20.667 (min: 7402.016 / max: 7488.832) us

stack fftn
get_fft_plan        :    CPU:  149.860 us   +/- 1.884 (min:  148.328 / max:  156.269) us     GPU-0:  155.638 us   +/- 2.036 (min:  153.984 / max:  162.336) us
fftn                :    CPU:   71.850 us   +/- 5.544 (min:   67.287 / max:   92.867) us     GPU-0:17079.713 us   +/-32.191 (min:16997.055 / max:17136.192) us

stack fft
get_fft_plan        :    CPU:  151.555 us   +/- 1.660 (min:  150.010 / max:  156.620) us     GPU-0:  157.618 us   +/- 1.875 (min:  155.744 / max:  163.008) us
fft                 :    CPU:   16.200 us   +/- 1.147 (min:   14.933 / max:   20.014) us     GPU-0: 2073.898 us   +/- 4.135 (min: 2066.368 / max: 2080.992) us
```

Script:
```python
import cupy as cp
from cupyx.time import repeat
from cupyx.scipy.fftpack import get_fft_plan

a = cp.random.random((256, 512, 512), dtype=cp.float32) + 1j* cp.random.random((256, 512, 512), dtype=cp.float32)
assert a.dtype == cp.complex64

print("full fftn")
print(repeat(get_fft_plan, (a,), n_repeat=20))
print(repeat(cp.fft.fftn, (a,), n_repeat=20))
print()

print("stack fftn")
print(repeat(get_fft_plan, (a, None, (-1, -2)), n_repeat=20))
print(repeat(cp.fft.fftn, (a, (-1, -2)), n_repeat=20))
print()

print("stack fft")
print(repeat(get_fft_plan, (a, None, (-1,)), n_repeat=20))
print(repeat(cp.fft.fft, (a,), n_repeat=20))
print()
```

**Comment by leofang** [comment]
https://github.com/cupy/cupy/issues/4746#issuecomment-785543235 is fixed in 5ad142f.

Jenkins, test this please

**Comment by leofang** [comment]
Looks like if I mark it slow (f3523ab), all tests in `tests/cupy_tests/fft_tests/test_callback.py` do not run. @emcastillo Is this OK? Should I undo it? (The entire file takes about 3 mins to run, so not marking it slow is perhaps fine?)

**Comment by leofang** [comment]
> Looks like if I mark it slow ([f3523ab](https://github.com/cupy/cupy/commit/f3523ab95d445ad88240e6f9859ee59b4940e7db)), all tests in `tests/cupy_tests/fft_tests/test_callback.py` do not run. @emcastillo Is this OK? Should I undo it? (The entire file takes about 3 mins to run, so not marking it slow is perhaps fine?)

I reverted it. 

Jenkins, test this please

File: cupy/cuda/cufft.pyx [hunk 1]
```diff
@@ -35,9 +34,13 @@ cdef enum:
 
 
 cdef extern from 'cupy_cufft.h' nogil:
+    # we duplicate some types here to avoid cimporting from driver/runtime,
+    # as we don't include their .pxd files in the sdist
+    ctypedef void* Stream 'cudaStream_t'
+    ctypedef int DataType 'cudaDataType'
```
Comment (emcastillo) [inline cupy/cuda/cufft.pyx:40]: These are not going to be changed anytime soon so it should be ok,
but I don't like this. However, I understand there is no other way of dealing with this.

**Review by emcastillo — APPROVED** [review]
In general the fix LGTM and I am glad to see that there was no performance penalty.
Let's wait @kmaehashi to do a pass over it :)

**Review by kmaehashi — APPROVED** [review]
LGTM!

**Comment by leofang** [comment]
Thanks @kmaehashi @emcastillo! Sorry for the all the troubles with the fft callbacks...Hope this time we fix it for good!

**Pull request merged**